developer_docs/order_docs.md

```diff
@@ -0,0 +1,427 @@
+import numpy as np
+
+if __name__ == "__main__":
+  a = np.array([[1, 2, 3],
+                [4, 5, 6],
+                [7, 8, 9]], order='C') # order='C' is the default in numpy.array
+  print(a.ravel('K'))
+ ```  
+ 
+This python script will output `[1 2 3 4 5 6 7 8 9]`, notice that the rows are stored one after the other.
+
+### Printing and indexing in `Numpy`
+
+`order` in `Numpy` does not affect the indexing or the printing, unlike `transposing`, the `shape` of the array remains the same, only the `strides` change, example:  
+
+```python
+import numpy as np
+
+if __name__ == "__main__":
+   a = np.array([[1, 2],
+                [4, 5],
+                [7, 8]], order='F')
+   b = np.array([[1, 2],
+                [4, 5],
+                [7, 8]], order='C')
+   print(a.shape, a.strides) # output: (3, 2) (8, 24)
+   print(b.shape, b.strides) # output: (3, 2) (16, 8)
+   print(a)
+   # output:[[1 2]
+   #         [4 5]
+   #         [7 8]] 
+   print(b)
+   # output:[[1 2]
+   #         [4 5]
+   #         [7 8]]
+   
+   print(a[2][1], a[0][0], a[1]) # output: 8 1 [4 5]
+   print(b[2][1], b[0][0], b[1]) # output: 8 1 [4 5]
+```  
+
+`arr.strides` is the variable that helps us navigate the array (indexing/printing), it tells us how many bytes we have to skip in memory to move to the next position along a certain axis (dimension). For example for `memory_layout_of_a = [1 4 7 2 5 8]` and `strides_of_a = (8, 24)`, we have to skip 8 bytes (1 element for `int64`) to move to the next row, but 24 bytes (3 elements for `int64`) to get to the same position in the next column of `a`.  
+`a[2][1]` would give us `'8'`, using the `strides`: `2 * 8 + 1 * 24 = 40`, which means that in the flattened array, we would have to skip `40` bytes to get the value of `a[2][1]`, each element is 8 bytes, so we would have to skip `40 / 8 = 5` elements to get to `'8'`
+
+## Pyccel's C code
+
+In `Pyccel`'s C code, we try to clone `Numpy`'s indexing/printing and memory layout conventions.
+
+### Ordering in C code
+
+The arrays in `C` code, are flattened into a one dimensional string, `strides` and `shape` are used to navigate the array (unlike `Numpy`. `Pyccel`'s strides use 'number of elements' instead of 'number of bytes' as a unit)
+While the `order_c ndarrays` only require a simple copy to be populated, `order_f` array creation requires slightly different steps.  
+
+Example:  
+
+  `order_c` creation  
+   &nbsp;&nbsp;&nbsp;&nbsp;1. allocate/create `order_c ndarray`  
+   &nbsp;&nbsp;&nbsp;&nbsp;2. copy values to `ndarray`  
+
+  `order_f` creation  
+   &nbsp;&nbsp;&nbsp;&nbsp;1. allocate/create temporary `order_c ndarray`  
+   &nbsp;&nbsp;&nbsp;&nbsp;2. copy values to temporary `ndarray`  
+   &nbsp;&nbsp;&nbsp;&nbsp;3. allocate/create `order_f ndarray`  
+   &nbsp;&nbsp;&nbsp;&nbsp;4. copy temporary `ndarray` elements to final `order_f ndarray` using `strides` and `shape`, this will create a column-major version of the temporary `order_c ndarray`  
```

Comment:
I think this step is only needed for the `np.array` function. As `np.array` is not the only way to populate an array (e.g. `np.ones`), I think this statement is a bit misleading

Reply:
Oh, I only had `np.array` in mind when writing this!
This means I might have to make more changes.

Reply:
Maybe just make it clear what you are talking about ?

Example:
 To translate the following:
```python
a = np.array([[1, 2, 3], [4, 5, 6]], order=?)
```

**`order_c`**
1.    allocate/create `order_c ndarray`
2.    copy values to `ndarray`

**`order_f`**
1.    allocate/create temporary `order_c ndarray` (contains the list)
2.    copy values to temporary `ndarray` (initialises the list)
3.    allocate/create `order_f ndarray`
4.    copy temporary `ndarray` elements to final `order_f ndarray` using `strides` and `shape`, this will create a column-major version of the temporary `order_c ndarray`

Reply:
@EmilyBourne, @bauom, are there any other functions that change behaviour depending on the order?
If that's the case then we might need to talk about them in here?
For now, this talks generally about order, then goes into the specifics of array creation depending on the order (`np.array`).

Reply:
Nothing comes to mind apart from copying data. But that is the same case as here. I'm not sure I would even classify `array` as a function changing its behaviour depending on the order. Both orders do:
- Allocate memory
- Copy data to memory

It's just that for the second step we are copying a C-ordered array to a F-ordered array so there needs to be some reorganisation